### PR TITLE
[Large 2D load for GEMM - #5] Use the maximum 2D load capacity as possible to load the dot operands.

### PR DESCRIPTION
**Pull request by chengjunlu**
Support the repCluster field in 2D load ops lowering and use the maximum 2D load capacity as possible.

### CODE REVIEW

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 1]
```diff
@@ -377,25 +379,35 @@ struct LoadOpConversion
...
-        isOperandA ? dpasLayout.getShapeA() : dpasLayout.getShapeB();
-    SmallVector<int64_t> elemsPerInstr = {operandShape[0], operandShape[1]};
-    int64_t elemsPerLane = product<int64_t>(elemsPerInstr) /
-                           product<unsigned>(getThreadsPerWarp(dpasLayout));
+    SmallVector<unsigned> dpasInstShape = isOperandA
+                                              ? dpasLayout.getDPASInstShapeA()
+                                              : dpasLayout.getDPASInstShapeB();
+    SmallVector<int64_t> elemsPerDPASInst = {dpasInstShape[0],
+                                             dpasInstShape[1]};
+    int64_t elemsPerLanePerDPASInst =
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:387]: int64_t -> unsigned

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 2]
```diff
@@ -377,25 +379,35 @@ struct LoadOpConversion
...
     Type elemType = (isOperandA && eltTy != f32_ty) ? i16_ty : i32_ty;
     unsigned opsPerChannel = dpasLayout.getOpsPerChannel();
-    elemsPerLane = isOperandA ? elemsPerLane / (opsPerChannel == 4 ? 2 : 1)
-                              : elemsPerLane / opsPerChannel;
-    Type load2DGenXType = LLVM::getFixedVectorType(elemType, elemsPerLane);
-
-    // Outer dim for A is the M, for B is the N. Inner dim for both is the K.
-    int outerDimWarpNum = std::min<int>(
-        warpsPerCTA[opIdx], ceil(tensorShape[opIdx], elemsPerInstr[opIdx]));
+    int64_t packedElemsPerLanePerDPASInst =
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:396]: use unsigned ?

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 3]
```diff
@@ -377,25 +379,35 @@ struct LoadOpConversion
...
+        LLVM::getFixedVectorType(elemType, packedElemsPerLanePerDPASInst);
+
+    // Outer dim: Dim M or N. Inner dim: Dim K.
+    // Round the warp id fit into the tensor shape.
+    auto repCluster = dpasLayout.getRepCluster();
+    SmallVector<unsigned> warpShape =
+        isOperandA ? dpasLayout.getShapeA() : dpasLayout.getShapeB();
+    unsigned outerDimRequiredWarpNum =
+        mlir::ceil<unsigned>(tensorShape[opIdx], warpShape[opIdx]);
+    int outerDimWarpNum =
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:409]: use unsigned

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 4]
```diff
@@ -407,65 +419,153 @@ struct LoadOpConversion
...
+            return failure();
+          }
+
+          unsigned packedRowNum =
+              opIdx == 0 ? numOperandsOuterDimPerLoad : numOperandsKDimPerLoad;
+          unsigned packedColNum =
+              opIdx == 0 ? numOperandsKDimPerLoad : numOperandsOuterDimPerLoad;
+          unsigned offset = 0;
+          // The register value returned by 2D load is contiguous on the row.
+          for (int col = 0; col < packedColNum; col++) {
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:525]: use prefix increment

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 5]
```diff
@@ -407,65 +419,153 @@ struct LoadOpConversion
...
+              /*tile_height*/ tileHeight,
+              /*v_blocks*/ vBlocks,
+              /*transpose*/ false,
+              /*vnni_transform*/
+              (!isOperandA && eltTy.getIntOrFloatBitWidth() != 32));
+          if (failed(load2dOp.verify())) {
+            // Explicitly invoke verifier because `triton_gen` ops are
+            // immediately lowered further to a builtin call.
+            return failure();
+          }
```
Comment (victor-eds) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:517]: I assume we cannot check some kind of preconditions to early exit. If not, LGTM.

**Review by victor-eds — APPROVED** [review]
LGTM

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp (hunk 5 above)
Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:517]: We can change this in another PR. Need to split the replacing stage into two: TritonGPU -> LLVM/TritonGen and then TritonGen -> LLVM.

**Review by whitneywhtsang — APPROVED** [review]

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 6]
```diff
@@ -404,68 +416,156 @@ struct LoadOpConversion
...
+          unsigned packedColNum =
+              opIdx == 0 ? numOperandsKDimPerLoad : numOperandsOuterDimPerLoad;
+          unsigned offset = 0;
+          // The register value returned by 2D load is contiguous on the row.
+          for (int col = 0; col < packedColNum; ++col) {
+            for (int row = 0; row < packedRowNum; ++row) {
+
+              Value loadVal = undef(packedDPASOperandType);
+              for (int elemIdx = 0; elemIdx < packedElemsPerLanePerDPASInst;
+                   elemIdx++) {
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:530]:
```suggestion
                   ++elemIdx) {
```

**Review by etiotto — APPROVED** [review]

**Pull request merged**